src/tunnel.cc

```diff
@@ -250,75 +269,110 @@ class TunnelStateData: public PeerSelectionInitiator
 
     void copyClientBytes();
     void copyServerBytes();
+
+    /// client-to-Squid connection closure cleanup; may destroy us
+    void handleClientClosure();
+
+    /// non-retriable Squid-to-server connection closure cleanup; may destroy us
+    void handleServerClosure();
 };
 
 static ERCB tunnelErrorComplete;
-static CLCB tunnelServerClosed;
 static CLCB tunnelClientClosed;
 static CTCB tunnelTimeout;
 static EVH tunnelDelayedClientRead;
 static EVH tunnelDelayedServerRead;
 
-static void
-tunnelServerClosed(const CommCloseCbParams &params)
+/// destroys the tunnel (after performing potentially-throwing cleanup)
+void
+TunnelStateData::deleteThis()
 {
-    TunnelStateData *tunnelState = (TunnelStateData *)params.data;
-    debugs(26, 3, HERE << tunnelState->server.conn);
-    tunnelState->server.conn = NULL;
-    tunnelState->server.writer = NULL;
-
-    if (tunnelState->request != NULL)
-        tunnelState->request->hier.stopPeerClock(false);
-
-    if (tunnelState->noConnections()) {
-        // ConnStateData pipeline should contain the CONNECT we are performing
-        // but it may be invalid already (bug 4392)
-        if (tunnelState->http.valid() && tunnelState->http->getConn()) {
-            auto ctx = tunnelState->http->getConn()->pipeline.front();
-            if (ctx != nullptr)
+    assert(noConnections());
+    // ConnStateData pipeline should contain the CONNECT we are performing
+    // but it may be invalid already (bug 4392)
+    if (const auto h = http.valid()) {
+        if (const auto c = h->getConn())
+            if (const auto ctx = c->pipeline.front())
                 ctx->finished();
-        }
-        delete tunnelState;
-        return;
     }
+    delete this;
+}
 
-    if (!tunnelState->client.writer) {
-        tunnelState->client.conn->close();
-        return;
-    }
+void
+TunnelStateData::handleServerClosure()
+{
+    server.conn = nullptr;
+    server.writer = nullptr;
+
+    if (request)
+        request->hier.stopPeerClock(false);
+
+    if (noConnections())
+        return deleteThis();
+
+    if (!client.writer)
+        client.conn->close();
 }
 
+void
+TunnelStateData::handleClientClosure()
+{
+    client.conn = nullptr;
+    client.writer = nullptr;
+
+    if (noConnections())
+        return deleteThis();
+
+    if (!server.writer)
+        server.conn->close();
+}
+
+/// handles closures of the Squid-to-server connection
 static void
-tunnelClientClosed(const CommCloseCbParams &params)
+tunnelServerClosed(const CommCloseCbParams &params)
 {
-    TunnelStateData *tunnelState = (TunnelStateData *)params.data;
-    debugs(26, 3, HERE << tunnelState->client.conn);
-    tunnelState->client.conn = NULL;
-    tunnelState->client.writer = NULL;
+    const auto tunnelState = reinterpret_cast<TunnelStateData *>(params.data);
```

Comment:
It looks like the guts of this rewritten function (i.e. everything below the reinterpret_cast) should become TunnelStateData::serverClosed(). There is a similar method in FwdState already.

Reply:
Done.